src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -149,15 +176,355 @@
+      .getFactor().doubleValue());
+  }
+
+  /**
+   * Build new {@link Voucher} based on processed {@link Invoice}
+   *
+   * @param invoice invoice {@link Invoice} to be approved
+   * @return completable future with {@link Voucher}
+   */
+  private CompletableFuture<Voucher> buildNewVoucher(Invoice invoice) {
+
+    Voucher voucher = new Voucher();
+    voucher.setInvoiceId(invoice.getId());
+    voucher.setInvoiceCurrency(invoice.getCurrency());
+    setDefaultRequiredFields(voucher);
+
+    return getSystemCurrency()
+      .thenCompose(systemCurrency -> getExchangeRate(voucher.withSystemCurrency(systemCurrency)))
+      .thenAccept(voucher::setExchangeRate)
+      .thenCompose(v -> voucherHelper.generateVoucherNumber())
+      .thenApply(voucher::withVoucherNumber);
+  }
+
+  /**
+   *  Retrieves systemCurrency from mod-configuration
+   *  <ul>
+   *    <li>if config is empty than use {@link #DEFAULT_SYSTEM_CURRENCY}</li>
+   *    <li>if an error is returned from the mod-configuration, stop transition to {@link Invoice.Status#APPROVED} with error</li>
+   *  </ul>
+   *
+   * @return
+   */
+  private CompletableFuture<String> getSystemCurrency() {
+    return loadConfiguration(SYSTEM_CONFIG_NAME, LOCALE_SETTINGS)
+      .thenApply(configs -> {
+        JsonObject config = configs.stream().map(conf -> new JsonObject(conf.getValue())).findFirst().orElseGet(JsonObject::new);
+        return config.getString(SYSTEM_CURRENCY_PROPERTY_NAME, DEFAULT_SYSTEM_CURRENCY);
+      });
+  }
+
+  /**
+   *  Handles creation (or update) of prepared voucher and voucher lines creation
+   *
+   * @param invoiceLines {@link List<InvoiceLine>} associated with processed invoice
+   * @param voucher associated with processed invoice
+   * @return CompletableFuture that indicates when handling is completed
+   */
+  private CompletableFuture<Void> handleVoucherWithLines(List<InvoiceLine> invoiceLines, Voucher voucher) {
+    return groupFundDistributionsByExternalAccountNo(invoiceLines)
+      .thenCompose(fundDistributionsGroupedByExternalAccountNo -> {
+        List<VoucherLine> voucherLines = buildVoucherLineRecords(fundDistributionsGroupedByExternalAccountNo, invoiceLines,
+            voucher);
+        Double calculatedAmount = HelperUtils.calculateVoucherAmount(voucher, voucherLines);
+        voucher.setAmount(calculatedAmount);
+        return handleVoucher(voucher)
+          .thenAccept(voucherWithId -> populateVoucherId(voucherLines, voucher))
+          .thenCompose(v -> createVoucherLinesRecords(voucherLines));
+      });
+  }
+
+  /**
+   * Prepares the data necessary for the generation of voucher lines based on the invoice lines found
+   *
+   * @param invoiceLines {@link List<InvoiceLine>} associated with processed {@link Invoice}
+   * @return {@link InvoiceLine#fundDistributions} grouped by {@link Fund#externalAccountNo}
+   */
+  private CompletableFuture<Map<String, List<FundDistribution>>> groupFundDistributionsByExternalAccountNo(
```

Comment:
Wow that's a long method name!  Not a huge deal but might make sense to abbreviate... 

`groupFundDistroByExternalAccountNo` or even `groupFundDistroByExternalAcctNo`

Reply:
Shortened it.